File: src/core/timeline/state.py
"""Timeline state management."""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Dict, List, Optional, Set, Any
from uuid import UUID

from pydantic import BaseModel, Field

from .models import TimelineTrack, TimelineSegment, PlaybackState, TrackType


class SelectionState(BaseModel):
    """State for segment selection."""
    selected_segments: Set[UUID] = Field(default_factory=set, description="IDs of selected segments")
    selected_tracks: Set[UUID] = Field(default_factory=set, description="IDs of selected tracks")
    selection_start_time: Optional[float] = Field(default=None, description="Start time of selection range")
    selection_end_time: Optional[float] = Field(default=None, description="End time of selection range")

    def is_segment_selected(self, segment_id: UUID) -> bool:
        """Check if a segment is selected."""
        return segment_id in self.selected_segments

    def is_track_selected(self, track_id: UUID) -> bool:
        """Check if a track is selected."""
        return track_id in self.selected_tracks

    def clear_selection(self) -> None:
        """Clear all selections."""
        self.selected_segments.clear()
        self.selected_tracks.clear()
        self.selection_start_time = None
        self.selection_end_time = None

    def select_segment(self, segment_id: UUID) -> None:
        """Select a single segment."""
        self.selected_segments.clear()
        self.selected_tracks.clear()
        self.selected_segments.add(segment_id)

    def select_track(self, track_id: UUID) -> None:
        """Select a track and all its segments."""
        self.selected_tracks.clear()
        self.selected_segments.clear()
        self.selected_tracks.add(track_id)

    def select_time_range(self, start_time: float, end_time: float) -> None:
        """Select all segments in a time range."""
        self.selected_segments.clear()
        self.selected_tracks.clear()
        self.selection_start_time = start_time
        self.selection_end_time = end_time


class PlaybackStateModel(BaseModel):
    """State for video playback."""
    state: PlaybackState = Field(default=PlaybackState.STOPPED, description="Current playback state")
    current_time: float = Field(default=0.0, ge=0, description="Current playback time in seconds")
    playback_rate: float = Field(default=1.0, ge=0.1, le=4.0, description="Playback speed multiplier")
    loop_enabled: bool = Field(default=False, description="Whether playback loops")
    in_point: Optional[float] = Field(default=None, ge=0, description="Playback start point")
    out_point: Optional[float] = Field(default=None, ge=0, description="Playback end point")

    def play(self) -> None:
        """Set state to playing."""
        self.state = PlaybackState.PLAYING

    def pause(self) -> None:
        """Set state to paused."""
        self.state = PlaybackState.PAUSED

    def stop(self) -> None:
        """Set state to stopped and reset current time."""
        self.state = PlaybackState.STOPPED
        self.current_time = 0.0

    def scrub_to(self, time: float) -> None:
        """Scrub to a specific time."""
        self.state = PlaybackState.SCRUBBING
        self.current_time = max(0, time)

    def set_in_point(self, time: float) -> None:
        """Set the in point for playback."""
        self.in_point = time
        if self.out_point and time >= self.out_point:
            raise ValueError("In point must be before out point")

    def set_out_point(self, time: float) -> None:
        """Set the out point for playback."""
        self.out_point = time
        if self.in_point and time <= self.in_point:
            raise ValueError("Out point must be after in point")


class ViewportState(BaseModel):
    """State for timeline viewport."""
    zoom_level: float = Field(default=1.0, gt=0.1, le=10.0, description="Timeline zoom level")
    scroll_position: float = Field(default=0.0, ge=0, description="Horizontal scroll position")
    track_heights: Dict[UUID, int] = Field(default_factory=dict, description="Custom track heights")
    collapsed_tracks: Set[UUID] = Field(default_factory=set, description="IDs of collapsed tracks")
    show_waveforms: bool = Field(default=True, description="Whether to show audio waveforms")
    show_grid: bool = Field(default=True, description="Whether to show timeline grid")
    snap_to_grid: bool = Field(default=True, description="Whether to snap to grid")
    grid_size: float = Field(default=0.1, gt=0, description="Grid size in seconds")

    def zoom_in(self, factor: float = 1.2) -> None:
        """Zoom in the timeline."""
        self.zoom_level = min(self.zoom_level * factor, 10.0)

    def zoom_out(self, factor: float = 1.2) -> None:
        """Zoom out the timeline."""
        self.zoom_level = max(self.zoom_level / factor, 0.1)

    def scroll_to_time(self, time: float) -> None:
        """Scroll to show the given time."""
        self.scroll_position = time

    def toggle_track_collapsed(self, track_id: UUID) -> None:
        """Toggle track collapsed state."""
        if track_id in self.collapsed_tracks:
            self.collapsed_tracks.remove(track_id)
        else:
            self.collapsed_tracks.add(track_id)

    def is_track_collapsed(self, track_id: UUID) -> bool:
        """Check if a track is collapsed."""
        return track_id in self.collapsed_tracks


class UndoRedoState(BaseModel):
    """State for undo/redo functionality."""
    can_undo: bool = Field(default=False, description="Whether undo is available")
    can_redo: bool = Field(default=False, description="Whether redo is available")
    undo_stack: List[Dict[str, Any]] = Field(default_factory=list, description="Undo stack")
    redo_stack: List[Dict[str, Any]] = Field(default_factory=list, description="Redo stack")
    max_history_size: int = Field(default=50, description="Maximum history size")

    def push_state(self, state_data: Dict[str, Any]) -> None:
        """Push a new state to the undo stack."""
        self.undo_stack.append(state_data)
        self.redo_stack.clear()  # Clear redo stack when new action is performed

        # Limit history size
        if len(self.undo_stack) > self.max_history_size:
            self.undo_stack.pop(0)

        self.can_undo = True
        self.can_redo = False

    def undo(self) -> Optional[Dict[str, Any]]:
        """Undo the last action."""
        if not self.undo_stack:
            return None

        state_data = self.undo_stack.pop()
        self.redo_stack.append(state_data)

        self.can_undo = len(self.undo_stack) > 0
        self.can_redo = True

        return state_data

    def redo(self) -> Optional[Dict[str, Any]]:
        """Redo the last undone action."""
        if not self.redo_stack:
            return None

        state_data = self.redo_stack.pop()
        self.undo_stack.append(state_data)

        self.can_undo = True
        self.can_redo = len(self.redo_stack) > 0

        return state_data




@dataclass
class TimelineEvent:
    """Event for timeline state changes."""
    event_type: str
    timestamp: datetime = field(default_factory=datetime.utcnow)
    user_id: Optional[str] = None
    data: Dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        """Convert event to dictionary."""
        return {
            "event_type": self.event_type,
            "timestamp": self.timestamp.isoformat(),
            "user_id": self.user_id,
            "data": self.data,
        }


class TimelineEventBus:
    """Event bus for timeline state changes."""

    def __init__(self):
        self.listeners: Dict[str, List[callable]] = {}

    def subscribe(self, event_type: str, listener: callable) -> None:
        """Subscribe to an event type."""
        if event_type not in self.listeners:
            self.listeners[event_type] = []
        self.listeners[event_type].append(listener)

    def unsubscribe(self, event_type: str, listener: callable) -> None:
        """Unsubscribe from an event type."""
        if event_type in self.listeners:
            try:
                self.listeners[event_type].remove(listener)
            except ValueError:
                pass

    def emit(self, event: TimelineEvent) -> None:
        """Emit an event to all listeners."""
        if event.event_type in self.listeners:
            for listener in self.listeners[event.event_type]:
                try:
                    listener(event)
                except Exception as e:
                    # Log error but don't break other listeners
                    print(f"Error in event listener: {e}")


# Global event bus instance
timeline_event_bus = TimelineEventBus()


class TimelineStateManager:
    """Manager for timeline state coordination."""

    def __init__(self):
        self.selection_state = SelectionState()
        self.playback_state = PlaybackStateModel()
        self.viewport_state = ViewportState()
        self.undo_redo_state = UndoRedoState()
        self.event_bus = timeline_event_bus

    def get_current_time(self) -> float:
        """Get the current playback time."""
        return self.playback_state.current_time

    def set_current_time(self, time: float) -> None:
        """Set the current playback time."""
        self.playback_state.scrub_to(time)
        self.event_bus.emit(TimelineEvent(
            event_type="time_changed",
            data={"time": time}
        ))

    def select_segment(self, segment_id: UUID) -> None:
        """Select a segment."""
        self.selection_state.select_segment(segment_id)
        self.event_bus.emit(TimelineEvent(
            event_type="segment_selected",
            data={"segment_id": str(segment_id)}
        ))

    def select_track(self, track_id: UUID) -> None:
        """Select a track."""
        self.selection_state.select_track(track_id)
        self.event_bus.emit(TimelineEvent(
            event_type="track_selected",
            data={"track_id": str(track_id)}
        ))

    def clear_selection(self) -> None:
        """Clear all selections."""
        self.selection_state.clear_selection()
        self.event_bus.emit(TimelineEvent(event_type="selection_cleared"))

    def play(self) -> None:
        """Start playback."""
        self.playback_state.play()
        self.event_bus.emit(TimelineEvent(event_type="playback_started"))

    def pause(self) -> None:
        """Pause playback."""
        self.playback_state.pause()
        self.event_bus.emit(TimelineEvent(event_type="playback_paused"))

    def stop(self) -> None:
        """Stop playback."""
        self.playback_state.stop()
        self.event_bus.emit(TimelineEvent(event_type="playback_stopped"))

    def zoom_in(self, factor: float = 1.2) -> None:
        """Zoom in the timeline."""
        self.viewport_state.zoom_in(factor)
        self.event_bus.emit(TimelineEvent(
            event_type="viewport_changed",
            data={"zoom_level": self.viewport_state.zoom_level}
        ))

    def zoom_out(self, factor: float = 1.2) -> None:
        """Zoom out the timeline."""
        self.viewport_state.zoom_out(factor)
        self.event_bus.emit(TimelineEvent(
            event_type="viewport_changed",
            data={"zoom_level": self.viewport_state.zoom_level}
        ))

    def can_undo(self) -> bool:
        """Check if undo is available."""
        return self.undo_redo_state.can_undo

    def can_redo(self) -> bool:
        """Check if redo is available."""
        return self.undo_redo_state.can_redo

    def undo(self) -> Optional[Dict[str, Any]]:
        """Undo the last action."""
        state_data = self.undo_redo_state.undo()
        if state_data:
            self.event_bus.emit(TimelineEvent(
                event_type="state_undone",
                data=state_data
            ))
        return state_data

    def redo(self) -> Optional[Dict[str, Any]]:
        """Redo the last undone action."""
        state_data = self.undo_redo_state.redo()
        if state_data:
            self.event_bus.emit(TimelineEvent(
                event_type="state_redone",
                data=state_data
            ))
        return state_data

    def push_undo_state(self, state_data: Dict[str, Any]) -> None:
        """Push a new state to the undo stack."""
        self.undo_redo_state.push_state(state_data)
        self.event_bus.emit(TimelineEvent(
            event_type="state_pushed",
            data={"undo_available": True}
        ))

    def get_state_snapshot(self) -> Dict[str, Any]:
        """Get a snapshot of the current state."""
        return {
            "selection": self.selection_state.dict(),
            "playback": self.playback_state.dict(),
            "viewport": self.viewport_state.dict(),
            "undo_redo": self.undo_redo_state.dict(),
            "timestamp": datetime.utcnow().isoformat()
        }

    def restore_state(self, state_data: Dict[str, Any]) -> None:
        """Restore state from a snapshot."""
        # This would restore the state from the snapshot
        # Implementation depends on the specific state structure
        self.event_bus.emit(TimelineEvent(
            event_type="state_restored",
            data=state_data
        ))

File: src/core/timeline/state.ts
/**
 * Timeline state management system.
 *
 * This module provides comprehensive state management for timeline functionality
 * including playback, selection, viewport, and undo/redo operations.
 */

import { TimelineState, TimelineSegment, TimelineTrack } from './models';

export interface TimelineStateUpdate {
  type: 'segment' | 'track' | 'playback' | 'selection' | 'viewport';
  operation: 'add' | 'update' | 'delete' | 'move' | 'resize';
  targetId: string;
  data?: any;
  timestamp: number;
}

export interface TimelineEvent {
  type: string;
  payload: any;
  timestamp: number;
}

export class TimelineStateManager {
  private currentState: TimelineState;
  private history: TimelineState[] = [];
  private historyIndex: number = -1;
  private maxHistorySize: number = 50;
  private eventListeners: Map<string, Function[]> = new Map();
  private isUpdating: boolean = false;

  constructor(initialState: TimelineState) {
    this.currentState = JSON.parse(JSON.stringify(initialState));
    this.saveToHistory();
  }

  /**
   * Get the current timeline state.
   */
  getState(): TimelineState {
    return JSON.parse(JSON.stringify(this.currentState));
  }

  /**
   * Update the timeline state with validation.
   */
  async updateState(newState: TimelineState): Promise<void> {
    if (this.isUpdating) {
      throw new Error('State update already in progress');
    }

    this.isUpdating = true;

    try {
      // Validate state
      this.validateState(newState);

      // Store current state for undo
      const previousState = JSON.parse(JSON.stringify(this.currentState));

      // Update state
      this.currentState = JSON.parse(JSON.stringify(newState));

      // Save to history
      this.saveToHistory();

      // Emit state change event
      this.emitEvent('stateChanged', {
        previousState,
        newState: this.currentState,
        timestamp: Date.now()
      });

    } finally {
      this.isUpdating = false;
    }
  }

  /**
   * Add a segment to a track.
   */
  async addSegment(trackId: string, segment: TimelineSegment): Promise<void> {
    const newState = JSON.parse(JSON.stringify(this.currentState));

    const track = newState.tracks.find(t => t.id === trackId);
    if (!track) {
      throw new Error(`Track ${trackId} not found`);
    }

    // Validate segment doesn't overlap with existing segments
    this.validateSegmentOverlap(track.segments, segment);

    // Add segment
    track.segments.push(segment);

    // Sort segments by start time
    track.segments.sort((a, b) => a.startTime - b.startTime);

    await this.updateState(newState);
  }

  /**
   * Update a segment.
   */
  async updateSegment(segmentId: string, updates: Partial<TimelineSegment>): Promise<void> {
    const newState = JSON.parse(JSON.stringify(this.currentState));

    let updated = false;
    for (const track of newState.tracks) {
      const segmentIndex = track.segments.findIndex(s => s.id === segmentId);
      if (segmentIndex !== -1) {
        // Validate update doesn't create overlaps
        const updatedSegment = { ...track.segments[segmentIndex], ...updates };
        const otherSegments = track.segments.filter(s => s.id !== segmentId);
        this.validateSegmentOverlap(otherSegments, updatedSegment);

        // Apply updates
        track.segments[segmentIndex] = updatedSegment;
        updated = true;
        break;
      }
    }

    if (!updated) {
      throw new Error(`Segment ${segmentId} not found`);
    }

    await this.updateState(newState);
  }

  /**
   * Delete a segment.
   */
  async deleteSegment(segmentId: string): Promise<void> {
    const newState = JSON.parse(JSON.stringify(this.currentState));

    let deleted = false;
    for (const track of newState.tracks) {
      const segmentIndex = track.segments.findIndex(s => s.id === segmentId);
      if (segmentIndex !== -1) {
        track.segments.splice(segmentIndex, 1);
        deleted = true;
        break;
      }
    }

    if (!deleted) {
      throw new Error(`Segment ${segmentId} not found`);
    }

    await this.updateState(newState);
  }

  /**
   * Move segments to a new time position.
   */
  async moveSegments(segmentIds: string[], newStartTime: number): Promise<void> {
    const newState = JSON.parse(JSON.stringify(this.currentState));

    // Calculate time offset
    const segments = this.getSegmentsByIds(segmentIds);
    if (segments.length === 0) {
      throw new Error('No segments found for move operation');
    }

    const minStartTime = Math.min(...segments.map(s => s.startTime));
    const timeOffset = newStartTime - minStartTime;

    // Update all segments
    for (const track of newState.tracks) {
      track.segments.forEach(segment => {
        if (segmentIds.includes(segment.id)) {
          segment.startTime += timeOffset;
          segment.endTime += timeOffset;
        }
      });
    }

    await this.updateState(newState);
  }

  /**
   * Resize a segment.
   */
  async resizeSegment(segmentId: string, newStartTime?: number, newEndTime?: number): Promise<void> {
    const updates: Partial<TimelineSegment> = {};

    if (newStartTime !== undefined) {
      updates.startTime = newStartTime;
    }

    if (newEndTime !== undefined) {
      updates.endTime = newEndTime;
      updates.duration = newEndTime - (updates.startTime ?? this.getSegmentById(segmentId)?.startTime ?? 0);
    }

    await this.updateSegment(segmentId, updates);
  }

  /**
   * Select segments.
   */
  async selectSegments(segmentIds: string[]): Promise<void> {
    const newState = JSON.parse(JSON.stringify(this.currentState));

    // Clear all selections first
    for (const track of newState.tracks) {
      track.segments.forEach(segment => {
        segment.selected = false;
      });
    }

    // Select specified segments
    for (const track of newState.tracks) {
      track.segments.forEach(segment => {
        if (segmentIds.includes(segment.id)) {
          segment.selected = true;
        }
      });
    }

    await this.updateState(newState);
  }

  /**
   * Update playback state.
   */
  async updatePlayback(isPlaying: boolean, currentTime?: number): Promise<void> {
    const newState = JSON.parse(JSON.stringify(this.currentState));

    newState.isPlaying = isPlaying;
    if (currentTime !== undefined) {
      newState.currentTime = currentTime;
    }

    await this.updateState(newState);
  }

  /**
   * Update viewport settings.
   */
  async updateViewport(startTime: number, endTime: number, zoom?: number): Promise<void> {
    const newState = JSON.parse(JSON.stringify(this.currentState));

    newState.viewportStart = startTime;
    newState.viewportEnd = endTime;
    if (zoom !== undefined) {
      newState.zoom = zoom;
    }

    await this.updateState(newState);
  }

  /**
   * Undo the last operation.
   */
  async undo(): Promise<TimelineState> {
    if (!this.canUndo()) {
      throw new Error('No operations to undo');
    }

    this.historyIndex--;
    this.currentState = JSON.parse(JSON.stringify(this.history[this.historyIndex]));

    this.emitEvent('stateChanged', {
      previousState: this.historyIndex > 0 ? this.history[this.historyIndex - 1] : null,
      newState: this.currentState,
      timestamp: Date.now(),
      operation: 'undo'
    });

    return this.currentState;
  }

  /**
   * Redo the last undone operation.
   */
  async redo(): Promise<TimelineState> {
    if (!this.canRedo()) {
      throw new Error('No operations to redo');
    }

    this.historyIndex++;
    this.currentState = JSON.parse(JSON.stringify(this.history[this.historyIndex]));

    this.emitEvent('stateChanged', {
      previousState: this.history[this.historyIndex - 1],
      newState: this.currentState,
      timestamp: Date.now(),
      operation: 'redo'
    });

    return this.currentState;
  }

  /**
   * Check if undo is available.
   */
  canUndo(): boolean {
    return this.historyIndex > 0;
  }

  /**
   * Check if redo is available.
   */
  canRedo(): boolean {
    return this.historyIndex < this.history.length - 1;
  }

  /**
   * Add an event listener.
   */
  addEventListener(eventType: string, listener: Function): void {
    if (!this.eventListeners.has(eventType)) {
      this.eventListeners.set(eventType, []);
    }
    this.eventListeners.get(eventType)!.push(listener);
  }

  /**
   * Remove an event listener.
   */
  removeEventListener(eventType: string, listener: Function): void {
    const listeners = this.eventListeners.get(eventType);
    if (listeners) {
      const index = listeners.indexOf(listener);
      if (index !== -1) {
        listeners.splice(index, 1);
      }
    }
  }

  /**
   * Get a segment by ID.
   */
  private getSegmentById(segmentId: string): TimelineSegment | null {
    for (const track of this.currentState.tracks) {
      const segment = track.segments.find(s => s.id === segmentId);
      if (segment) {
        return segment;
      }
    }
    return null;
  }

  /**
   * Get segments by IDs.
   */
  private getSegmentsByIds(segmentIds: string[]): TimelineSegment[] {
    const segments: TimelineSegment[] = [];
    for (const track of this.currentState.tracks) {
      track.segments.forEach(segment => {
        if (segmentIds.includes(segment.id)) {
          segments.push(segment);
        }
      });
    }
    return segments;
  }

  /**
   * Validate timeline state.
   */
  private validateState(state: TimelineState): void {
    if (!state.tracks || !Array.isArray(state.tracks)) {
      throw new Error('Invalid timeline state: tracks must be an array');
    }

    // Validate each track
    state.tracks.forEach(track => {
      if (!track.id || !track.segments) {
        throw new Error('Invalid track: missing id or segments');
      }

      // Validate segments
      track.segments.forEach(segment => {
        if (segment.startTime >= segment.endTime) {
          throw new Error(`Invalid segment ${segment.id}: start time must be before end time`);
        }
      });

      // Check for overlaps within track
      for (let i = 0; i < track.segments.length; i++) {
        for (let j = i + 1; j < track.segments.length; j++) {
          const seg1 = track.segments[i];
          const seg2 = track.segments[j];
          if (seg1.startTime < seg2.endTime && seg2.startTime < seg1.endTime) {
            throw new Error(`Overlapping segments detected: ${seg1.id} and ${seg2.id}`);
          }
        }
      }
    });
  }

  /**
   * Validate segment doesn't overlap with existing segments.
   */
  private validateSegmentOverlap(existingSegments: TimelineSegment[], newSegment: TimelineSegment): void {
    existingSegments.forEach(segment => {
      if (segment.id !== newSegment.id &&
          segment.startTime < newSegment.endTime &&
          newSegment.startTime < segment.endTime) {
        throw new Error(`Segment overlap detected: ${segment.id} and ${newSegment.id}`);
      }
    });
  }

  /**
   * Save current state to history.
   */
  private saveToHistory(): void {
    // Remove any states after current index (when new operation after undo)
    if (this.historyIndex < this.history.length - 1) {
      this.history = this.history.slice(0, this.historyIndex + 1);
    }

    // Add current state
    this.history.push(JSON.parse(JSON.stringify(this.currentState)));
    this.historyIndex++;

    // Limit history size
    if (this.history.length > this.maxHistorySize) {
      this.history = this.history.slice(-this.maxHistorySize);
      this.historyIndex = Math.min(this.historyIndex, this.maxHistorySize - 1);
    }
  }

  /**
   * Emit an event to all listeners.
   */
  private emitEvent(eventType: string, payload: any): void {
    const listeners = this.eventListeners.get(eventType);
    if (listeners) {
      listeners.forEach(listener => {
        try {
          listener(payload);
        } catch (error) {
          console.error(`Error in event listener for ${eventType}:`, error);
        }
      });
    }
  }
}

export default TimelineStateManager;